### Introduction to data and baseline for the project

Autoreload for making it faster to work with the notebook. Initialize Fetcher.

In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
from fetcher import Fetcher

fetch = Fetcher('data/database.sqlite')

Get all matches for an initial look into the data

In [3]:
matches = fetch.get_all_matches()

In [4]:
len(matches)

25979

In [5]:
matches[0]

{'id': 1,
 'country_id': 1,
 'league_id': 1,
 'season': '2008/2009',
 'stage': 1,
 'date': '2008-08-17 00:00:00',
 'match_api_id': 492473,
 'home_team_api_id': 9987,
 'away_team_api_id': 9993,
 'home_team_goal': 1,
 'away_team_goal': 1,
 'home_player_X1': None,
 'home_player_X2': None,
 'home_player_X3': None,
 'home_player_X4': None,
 'home_player_X5': None,
 'home_player_X6': None,
 'home_player_X7': None,
 'home_player_X8': None,
 'home_player_X9': None,
 'home_player_X10': None,
 'home_player_X11': None,
 'away_player_X1': None,
 'away_player_X2': None,
 'away_player_X3': None,
 'away_player_X4': None,
 'away_player_X5': None,
 'away_player_X6': None,
 'away_player_X7': None,
 'away_player_X8': None,
 'away_player_X9': None,
 'away_player_X10': None,
 'away_player_X11': None,
 'home_player_Y1': None,
 'home_player_Y2': None,
 'home_player_Y3': None,
 'home_player_Y4': None,
 'home_player_Y5': None,
 'home_player_Y6': None,
 'home_player_Y7': None,
 'home_player_Y8': None,
 'home_pl

In [6]:
leagues = fetch.get_leagues()

In [8]:
leagues

[{'id': 1, 'country_id': 1, 'name': 'Belgium'},
 {'id': 1729, 'country_id': 1729, 'name': 'England'},
 {'id': 4769, 'country_id': 4769, 'name': 'France'},
 {'id': 7809, 'country_id': 7809, 'name': 'Germany'},
 {'id': 10257, 'country_id': 10257, 'name': 'Italy'},
 {'id': 13274, 'country_id': 13274, 'name': 'Netherlands'},
 {'id': 15722, 'country_id': 15722, 'name': 'Poland'},
 {'id': 17642, 'country_id': 17642, 'name': 'Portugal'},
 {'id': 19694, 'country_id': 19694, 'name': 'Scotland'},
 {'id': 21518, 'country_id': 21518, 'name': 'Spain'},
 {'id': 24558, 'country_id': 24558, 'name': 'Switzerland'}]

Let's work with the English Premier League since it is the most popular league in Estonia

In [9]:
import pandas as pd

In [10]:
#It's easy to make a pandas dataframe from a list of dictionaries
matches_df = pd.DataFrame(matches)

In [11]:
matches_df.head()

B365A  B365D  B365H   BSA   BSD   BSH   BWA   BWD   BWH   GBA  ...    \
0   5.00   3.40   1.73  4.20  3.40  1.73  4.20  3.35  1.75  4.00  ...     
1   3.60   3.20   1.95  3.60  3.25  1.91  3.95  3.30  1.80  3.75  ...     
2   2.75   3.30   2.38  2.75  3.20  2.30  2.55  3.30  2.40  2.50  ...     
3   7.50   3.75   1.44  6.50  3.75  1.44  6.80  4.00  1.40  5.50  ...     
4   1.65   3.50   5.00  1.67  3.30  4.75  1.60  3.50  5.00  1.65  ...     

   home_team_api_id  home_team_goal  id  league_id  match_api_id  possession  \
0              9987               1   1          1        492473        None   
1             10000               0   2          1        492474        None   
2              9984               0   3          1        492475        None   
3              9991               5   4          1        492476        None   
4              7947               1   5          1        492477        None   

      season  shotoff  shoton  stage  
0  2008/2009     None    None      1  
1  2008/2009     None    None      1  
2  2008/2009     None    None      1  
3  2008/2009     None    None      1  
4  2008/2009     None    None      1  

[5 rows x 115 columns]

In [12]:
english = matches_df[matches_df['league_id']==1729]

In [15]:
english.describe()

B365A        B365D        B365H          BSA          BSD  \
count  3040.000000  3040.000000  3040.000000  1900.000000  1900.000000   
mean      4.910437     3.952720     2.701964     4.790621     3.773584   
std       3.909392     0.998305     1.689834     3.786069     0.837965   
min       1.220000     3.000000     1.100000     1.220000     3.000000   
25%       2.500000     3.300000     1.670000     2.500000     3.250000   
50%       3.500000     3.600000     2.200000     3.400000     3.400000   
75%       5.500000     4.200000     3.000000     5.500000     3.900000   
max      29.000000    11.000000    15.000000    26.000000     8.500000   

               BSH          BWA          BWD          BWH          GBA  \
count  1900.000000  3039.000000  3039.000000  3039.000000  1899.000000   
mean      2.625553     4.495166     3.803251     2.603906     4.566435   
std       1.655792     3.213833     0.882597     1.528913     3.301710   
min       1.100000     1.220000     2.900000     1.100000     1.250000   
25%       1.620000     2.400000     3.250000     1.650000     2.450000   
50%       2.100000     3.400000     3.400000     2.100000     3.400000   
75%       2.800000     5.250000     4.000000     2.875000     5.250000   
max      13.000000    21.000000     9.250000    12.500000    21.000000   

          ...      home_player_Y6  home_player_Y7  home_player_Y8  \
count     ...         3040.000000     3040.000000     3040.000000   
mean      ...            6.611184        6.722368        7.273026   
std       ...            0.602293        0.506515        0.515437   
min       ...            3.000000        5.000000        3.000000   
25%       ...            6.000000        6.000000        7.000000   
50%       ...            7.000000        7.000000        7.000000   
75%       ...            7.000000        7.000000        8.000000   
max       ...            8.000000        8.000000        9.000000   

       home_player_Y9  home_team_api_id  home_team_goal           id  \
count     3040.000000        3040.00000     3040.000000  3040.000000   
mean         7.670395        9142.27500        1.550987  3248.500000   
std          0.957768         753.08936        1.311615   877.716735   
min          6.000000        8191.00000        0.000000  1729.000000   
25%          7.000000        8528.00000        1.000000  2488.750000   
50%          7.000000        8659.00000        1.000000  3248.500000   
75%          8.000000        9879.00000        2.000000  4008.250000   
max         10.000000       10261.00000        9.000000  4768.000000   

       league_id  match_api_id       stage  
count     3040.0  3.040000e+03  3040.00000  
mean      1729.0  1.178767e+06    19.50000  
std          0.0  4.909627e+05    10.96766  
min       1729.0  4.890420e+05     1.00000  
25%       1729.0  7.946915e+05    10.00000  
50%       1729.0  1.127220e+06    19.50000  
75%       1729.0  1.537336e+06    29.00000  
max       1729.0  1.989079e+06    38.00000  

[8 rows x 105 columns]

As we can see, there are 3040 matches to work with. Interestingly, already the first few columns have some interesting facts. The mean of the odds that Bet365 offers for the home team is 2.7 but for the away team it is 4.9 which is quite a big difference.

Since different seasons have very different teams lining up, let's look at just one season for now.

In [18]:
english_2016 = english[english['season']=="2015/2016"]

In [20]:
english_2016.describe()

B365A       B365D       B365H  BSA  BSD  BSH         BWA  \
count  380.000000  380.000000  380.000000  0.0  0.0  0.0  379.000000   
mean     4.441421    3.901342    2.618816  NaN  NaN  NaN    4.085303   
std      2.924514    0.833620    1.290147  NaN  NaN  NaN    2.528689   
min      1.440000    3.100000    1.170000  NaN  NaN  NaN    1.440000   
25%      2.500000    3.400000    1.730000  NaN  NaN  NaN    2.450000   
50%      3.550000    3.600000    2.250000  NaN  NaN  NaN    3.250000   
75%      5.250000    4.000000    3.000000  NaN  NaN  NaN    4.750000   
max     17.000000    9.000000    8.000000  NaN  NaN  NaN   16.000000   

              BWD         BWH  GBA     ...      home_player_Y6  \
count  379.000000  379.000000  0.0     ...          380.000000   
mean     3.700607    2.525910  NaN     ...            6.284211   
std      0.694848    1.142504  NaN     ...            0.546771   
min      3.000000    1.160000  NaN     ...            3.000000   
25%      3.250000    1.720000  NaN     ...            6.000000   
50%      3.400000    2.250000  NaN     ...            6.000000   
75%      3.800000    2.900000  NaN     ...            7.000000   
max      7.000000    7.000000  NaN     ...            7.000000   

       home_player_Y7  home_player_Y8  home_player_Y9  home_team_api_id  \
count      380.000000      380.000000      380.000000        380.000000   
mean         6.426316        7.636842        7.931579       9211.450000   
std          0.596686        0.528562        0.737644        761.958195   
min          6.000000        6.000000        7.000000       8197.000000   
25%          6.000000        7.000000        8.000000       8557.500000   
50%          6.000000        8.000000        8.000000       8673.000000   
75%          7.000000        8.000000        8.000000       9888.250000   
max          8.000000        8.000000       10.000000      10261.000000   

       home_team_goal           id  league_id  match_api_id       stage  
count      380.000000   380.000000      380.0  3.800000e+02  380.000000  
mean         1.492105  4578.500000     1729.0  1.988773e+06   19.500000  
std          1.259242   109.840794        0.0  4.653757e+02   10.980313  
min          0.000000  4389.000000     1729.0  1.987032e+06    1.000000  
25%          1.000000  4483.750000     1729.0  1.988795e+06   10.000000  
50%          1.000000  4578.500000     1729.0  1.988890e+06   19.500000  
75%          2.000000  4673.250000     1729.0  1.988984e+06   29.000000  
max          6.000000  4768.000000     1729.0  1.989079e+06   38.000000  

[8 rows x 105 columns]

380 matches is a good sign since there are 20 teams, each playing 38 matches - we have to divide by 2 since there are two teams in each match.

In [29]:
#Order the matches by date and reset the index to make further processing easier
english_2016 = english_2016.sort_values('date').reset_index(drop=True)

In [31]:
english_2016.head()

B365A  B365D  B365H  BSA  BSD  BSH   BWA   BWD   BWH  GBA  ...    \
0   4.00    3.6   2.00  NaN  NaN  NaN  3.70  3.30  2.00  NaN  ...     
1  11.00    5.0   1.36  NaN  NaN  NaN  9.00  4.75  1.40  NaN  ...     
2   5.50    3.9   1.70  NaN  NaN  NaN  5.00  3.50  1.70  NaN  ...     
3   4.33    3.5   1.95  NaN  NaN  NaN  3.75  3.30  2.00  NaN  ...     
4   6.00    4.0   1.65  NaN  NaN  NaN  5.50  4.00  1.65  NaN  ...     

   home_team_api_id  home_team_goal    id  league_id  match_api_id  \
0              8678               0  4390       1729       1987033   
1              8455               2  4391       1729       1987034   
2              8668               2  4392       1729       1987035   
3              8197               4  4393       1729       1987036   
4             10260               1  4394       1729       1987037   

                                          possession     season  \
0  <possession><value><comment>58</comment><stats...  2015/2016   
1  <possession><value><comment>54</comment><stats...  2015/2016   
2  <possession><value><comment>65</comment><stats...  2015/2016   
3  <possession><value><comment>46</comment><stats...  2015/2016   
4  <possession><value><comment>46</comment><stats...  2015/2016   

                                             shotoff  \
0  <shotoff><value><stats><shotoff>1</shotoff></s...   
1  <shotoff><value><stats><shotoff>1</shotoff></s...   
2  <shotoff><value><stats><shotoff>1</shotoff></s...   
3  <shotoff><value><event_incident_typefk>743</ev...   
4  <shotoff><value><stats><shotoff>1</shotoff></s...   

                                              shoton  stage  
0  <shoton><value><stats><blocked>1</blocked></st...      1  
1  <shoton><value><stats><blocked>1</blocked></st...      1  
2  <shoton><value><stats><blocked>1</blocked></st...      1  
3  <shoton><value><stats><shoton>1</shoton></stat...      1  
4  <shoton><value><event_incident_typefk>876</eve...      1  

[5 rows x 115 columns]

Let's take one specific match to see what preprocessing we would have to do.

In [47]:
english_2016.loc[0]

B365A                                                               4
B365D                                                             3.6
B365H                                                               2
BSA                                                               NaN
BSD                                                               NaN
BSH                                                               NaN
BWA                                                               3.7
BWD                                                               3.3
BWH                                                                 2
GBA                                                               NaN
GBD                                                               NaN
GBH                                                               NaN
IWA                                                               3.3
IWD                                                               3.3
IWH                 

In [49]:
#don't have much to with NaNs
english_2016.loc[0].dropna()

B365A                                                               4
B365D                                                             3.6
B365H                                                               2
BWA                                                               3.7
BWD                                                               3.3
BWH                                                                 2
IWA                                                               3.3
IWD                                                               3.3
IWH                                                               2.1
LBA                                                                 4
LBD                                                               3.3
LBH                                                              2.05
PSA                                                              4.27
PSD                                                              3.65
PSH                 

We can see that there are some really messy attributes, let's see if we can make sense of them

In [46]:
english_2016.loc[0]['possession']

'<possession><value><comment>58</comment><stats><homepos>58</homepos><awaypos>42</awaypos></stats><event_incident_typefk>352</event_incident_typefk><elapsed>23</elapsed><subtype>possession</subtype><sortorder>0</sortorder><awaypos>42</awaypos><homepos>58</homepos><n>78</n><type>special</type><id>4653212</id></value><value><comment>58</comment><stats><homepos>58</homepos><awaypos>42</awaypos></stats><event_incident_typefk>352</event_incident_typefk><elapsed>44</elapsed><subtype>possession</subtype><sortorder>5</sortorder><awaypos>42</awaypos><homepos>58</homepos><n>199</n><type>special</type><id>4653579</id></value><value><comment>55</comment><stats><homepos>55</homepos><awaypos>45</awaypos></stats><event_incident_typefk>352</event_incident_typefk><elapsed>69</elapsed><subtype>possession</subtype><sortorder>4</sortorder><awaypos>45</awaypos><homepos>55</homepos><n>383</n><type>special</type><id>4654141</id></value><value><comment>57</comment><stats><homepos>57</homepos><awaypos>43</away

Looks like XML. Let's process it.

In [51]:
import xml.etree.ElementTree as ET
tree = ET.fromstring(english_2016.loc[0]['possession'])

In [63]:
for child in tree:
    print(child.tag,child.text)
    for second in child:
        print("\t",second.tag,second.text)
        for third in second:
            print("\t","\t",third.tag,third.text)

value None
	 comment 58
	 stats None
	 	 homepos 58
	 	 awaypos 42
	 event_incident_typefk 352
	 elapsed 23
	 subtype possession
	 sortorder 0
	 awaypos 42
	 homepos 58
	 n 78
	 type special
	 id 4653212
value None
	 comment 58
	 stats None
	 	 homepos 58
	 	 awaypos 42
	 event_incident_typefk 352
	 elapsed 44
	 subtype possession
	 sortorder 5
	 awaypos 42
	 homepos 58
	 n 199
	 type special
	 id 4653579
value None
	 comment 55
	 stats None
	 	 homepos 55
	 	 awaypos 45
	 event_incident_typefk 352
	 elapsed 69
	 subtype possession
	 sortorder 4
	 awaypos 45
	 homepos 55
	 n 383
	 type special
	 id 4654141
value None
	 comment 57
	 stats None
	 	 homepos 57
	 	 awaypos 43
	 event_incident_typefk 352
	 elapsed 89
	 subtype possession
	 sortorder 0
	 awaypos 43
	 homepos 57
	 n 490
	 type special
	 id 4654485


The "elapsed" seems to show how many minutes have elapsed from the start of the match and then homepos is a percentage of how much the home team has had the ball so far and awaypos the same for the away team. Possession can certainly be an important factor when training a model but not at all points of the match. It seems reasonable to extract the last possession number and use it during the training.

In [77]:
#find the last occurence of "homepos" and look a few characters before that for the number
string = english_2016.loc[0]['possession']
index = string.rfind('homepos')
homepos = string[index-4:index-2]
int(homepos)

57

In [79]:
newpos = []
for val in english_2016['possession']:
    index = val.rfind('homepos')
    homepos = val[index-4:index-2]
    newpos.append(int(homepos))

We only need the home team possession because the away team possession can easily be deduced from that and would provide no extra info for the neural net

In [81]:
english_2016['possession'] = newpos

It would probably make sense to have all the previous data when predicting the outcome of a match. However, for starters, let's say that the final 40 matches are our test data and all the previous matches are training data.

In [34]:
train = english_2016[:340]
test = english_2016[340:]